### PR TITLE
Fix colorization bug and add regression tests

### PR DESCRIPTION
**Pull request by pkmitre**
1) Fixes https://jira.oncprojectracking.org/browse/BONNIE-168

2) Adds rake task allowing colorization code to be regression tested

### CODE REVIEW

File: lib/tasks/colorization_tests.rake [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+    # Given a measure, population index, and an array of patients, calculate the patients against the
+    # population and return a matching array of results, including the specificsRationale which gets
+    # calculated on the front end (ie in a browser) for use in colorization; this makes use of PhantomJS to
+    # simulate a browser environment so we can run the front end code without using a browser.
+    def calculate_in_frontend(measure, population_index, patients)
+
+      results = nil
+
+      # Set up code for calculating with colorization results to be evaluated in PhantomJS; for this we write JS to a temp file
```
Comment (edeyoung) [inline lib/tasks/colorization_tests.rake]: For maintenance, it would be helpful to explain in more detail why the code needs to be written to an external temp file

File: lib/tasks/colorization_tests.rake [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+            if gold
+              if gold.result == result
+                passed += 1
+                print '.'
+              else
+                failed += 1
+                puts "\n\n  FAILURE: User #{measure.user.email} measure #{measure.cms_id} population #{population_index + 1} patient '#{patient.first} #{patient.last}'\n\n"
+                puts "             Difference:"
+                puts
+                puts hash_diff(result, gold.result).pretty_inspect.gsub(/^/, '               ')
```
Comment (edeyoung) [inline lib/tasks/colorization_tests.rake]: when you run it so you will get a difference, the output is a bit hard to parse. I think it would be more intuitive with less information. In the diff, it shows the particular criteria that make an element evaluate to true. I think I just care about which logic segment is evaluating differently, and then I'd go into the debugger view to find out the additional information.

Here's what I propose:

```
FAILURE: User epecqmtelligen@gmail.com measure CMS160v4 population 1 patient 'Thomas TESTT'

             Logic segments that calculated differently:

                 ["less_thn_4monthsEndsBeforeEndOf_B781F58B_C800_45AA_9052_C2B802EB7980_bec03ff0_88da_49ee_a2fd_21037ee0a3a3",
                   "prefix_0a41848e_2e58_4d1f_a550_49ad1af166ae_bec03ff0_88da_49ee_a2fd_21037ee0a3a3",
                   "SatisfiesAny_976700C3_C2C1_4B9E_BD13_474C16D9200C_F9240410_6E58_44B8_AA3A_C97DB4591690",
                   "SatisfiesAll_E463FEA2_9209_4FA0_AAEF_3466B24A0536_12733BC6_DBCD_4ABD_81D9_BA4775D473AF",
                   "Union_CC7491B7_6A9C_48A7_B828_7AB72D05B203_6F286505_F0D4_4E3F_B8BB_350D9A353988"]
```

Here's what it currently shows:

```
FAILURE: User epecqmtelligen@gmail.com measure CMS160v4 population 1 patient 'Thomas TESTT'

             Difference:

               {"rationale"=>
                 {"less_thn_4monthsEndsBeforeEndOf_B781F58B_C800_45AA_9052_C2B802EB7980_bec03ff0_88da_49ee_a2fd_21037ee0a3a3"=>
                   {"results"=>
                     [[{"id"=>"57169443b0ede6719f002939",
                        "json"=>
                         {"_id"=>"57169443b0ede6719f002939",
                          "admitTime"=>nil,
                          "admitType"=>nil,
                          "codes"=>{"CPT"=>["90834"]},
                          "description"=>"Encounter, Performed: Psych Visit",
                          "diagnosis"=>nil,
                          "dischargeDisposition"=>nil,
                          "dischargeTime"=>nil,
                          "end_time"=>1348152600,
                          "mood_code"=>"EVN",
                          "negationInd"=>nil,
                          "negationReason"=>nil,
                          "oid"=>"2.16.840.1.113883.3.560.1.79",
                          "performer_id"=>nil,
                          "principalDiagnosis"=>
                           {"_id"=>"5776af2eb0ede62f0f000003",
                            "code"=>"2506003",
                            "code_system"=>"SNOMED-CT",
                            "title"=>"Dysthymia"},
                          "reason"=>nil,
                          "specifics"=>nil,
                          "start_time"=>1348151400,
                          "status_code"=>{"HL7 ActStatus"=>["performed"]},
                          "time"=>nil}}],
                      [{"id"=>"57169443b0ede6719f002939",
                        "json"=>
                         {"_id"=>"57169443b0ede6719f002939",
                          "admitTime"=>nil,
                          "admitType"=>nil,
                          "codes"=>{"CPT"=>["90834"]},
                          "description"=>"Encounter, Performed: Psych Visit",
                          "diagnosis"=>nil,
                          "dischargeDisposition"=>nil,
                          "dischargeTime"=>nil,
                          "end_time"=>1348152600,
                          "mood_code"=>"EVN",
                          "negationInd"=>nil,
                          "negationReason"=>nil,
                          "oid"=>"2.16.840.1.113883.3.560.1.79",
                          "performer_id"=>nil,
                          "principalDiagnosis"=>
                           {"_id"=>"5776adedb0ede62ed3000060",
                            "code"=>"2506003",
                            "code_system"=>"SNOMED-CT",
                            "title"=>"Dysthymia"},
                          "reason"=>nil,
                          "specifics"=>nil,
                          "start_time"=>1348151400,
                          "status_code"=>{"HL7 ActStatus"=>["performed"]},
                          "time"=>nil}}]]},
                  "prefix_0a41848e_2e58_4d1f_a550_49ad1af166ae_bec03ff0_88da_49ee_a2fd_21037ee0a3a3"=>
                   {"results"=>
                     [[{"id"=>"57169443b0ede6719f002939",
                        "json"=>
                         {"_id"=>"57169443b0ede6719f002939",
                          "admitTime"=>nil,
                          "admitType"=>nil,
                          "codes"=>{"CPT"=>["90834"]},
                          "description"=>"Encounter, Performed: Psych Visit",
                          "diagnosis"=>nil,
                          "dischargeDisposition"=>nil,
                          "dischargeTime"=>nil,
                          "end_time"=>1348152600,
                          "mood_code"=>"EVN",
                          "negationInd"=>nil,
                          "negationReason"=>nil,
                          "oid"=>"2.16.840.1.113883.3.560.1.79",
                          "performer_id"=>nil,
                          "principalDiagnosis"=>
                           {"_id"=>"5776af2eb0ede62f0f000003",
                            "code"=>"2506003",
                            "code_system"=>"SNOMED-CT",
                            "title"=>"Dysthymia"},
                          "reason"=>nil,
                          "specifics"=>nil,
                          "start_time"=>1348151400,
                          "status_code"=>{"HL7 ActStatus"=>["performed"]},
                          "time"=>nil}}],
                      [{"id"=>"57169443b0ede6719f002939",
                        "json"=>
                         {"_id"=>"57169443b0ede6719f002939",
                          "admitTime"=>nil,
                          "admitType"=>nil,
                          "codes"=>{"CPT"=>["90834"]},
                          "description"=>"Encounter, Performed: Psych Visit",
                          "diagnosis"=>nil,
                          "dischargeDisposition"=>nil,
                          "dischargeTime"=>nil,
                          "end_time"=>1348152600,
                          "mood_code"=>"EVN",
                          "negationInd"=>nil,
                          "negationReason"=>nil,
                          "oid"=>"2.16.840.1.113883.3.560.1.79",
                          "performer_id"=>nil,
                          "principalDiagnosis"=>
                           {"_id"=>"5776adedb0ede62ed3000060",
                            "code"=>"2506003",
                            "code_system"=>"SNOMED-CT",
                            "title"=>"Dysthymia"},
                          "reason"=>nil,
                          "specifics"=>nil,
                          "start_time"=>1348151400,
                          "status_code"=>{"HL7 ActStatus"=>["performed"]},
                          "time"=>nil}}]]},
                  "SatisfiesAny_976700C3_C2C1_4B9E_BD13_474C16D9200C_F9240410_6E58_44B8_AA3A_C97DB4591690"=>
                   {"results"=>
                     [[{"id"=>"57169443b0ede6719f002939",
                        "json"=>
                         {"_id"=>"57169443b0ede6719f002939",
                          "admitTime"=>nil,
                          "admitType"=>nil,
                          "codes"=>{"CPT"=>["90834"]},
                          "description"=>"Encounter, Performed: Psych Visit",
                          "diagnosis"=>nil,
                          "dischargeDisposition"=>nil,
                          "dischargeTime"=>nil,
                          "end_time"=>1348152600,
                          "mood_code"=>"EVN",
                          "negationInd"=>nil,
                          "negationReason"=>nil,
                          "oid"=>"2.16.840.1.113883.3.560.1.79",
                          "performer_id"=>nil,
                          "principalDiagnosis"=>
                           {"_id"=>"5776af2eb0ede62f0f000003",
                            "code"=>"2506003",
                            "code_system"=>"SNOMED-CT",
                            "title"=>"Dysthymia"},
                          "reason"=>nil,
                          "specifics"=>nil,
                          "start_time"=>1348151400,
                          "status_code"=>{"HL7 ActStatus"=>["performed"]},
                          "time"=>nil}}],
                      [{"id"=>"57169443b0ede6719f002939",
                        "json"=>
                         {"_id"=>"57169443b0ede6719f002939",
                          "admitTime"=>nil,
                          "admitType"=>nil,
                          "codes"=>{"CPT"=>["90834"]},
                          "description"=>"Encounter, Performed: Psych Visit",
                          "diagnosis"=>nil,
                          "dischargeDisposition"=>nil,
                          "dischargeTime"=>nil,
                          "end_time"=>1348152600,
                          "mood_code"=>"EVN",
                          "negationInd"=>nil,
                          "negationReason"=>nil,
                          "oid"=>"2.16.840.1.113883.3.560.1.79",
                          "performer_id"=>nil,
                          "principalDiagnosis"=>
                           {"_id"=>"5776adedb0ede62ed3000060",
                            "code"=>"2506003",
                            "code_system"=>"SNOMED-CT",
                            "title"=>"Dysthymia"},
                          "reason"=>nil,
                          "specifics"=>nil,
                          "start_time"=>1348151400,
                          "status_code"=>{"HL7 ActStatus"=>["performed"]},
                          "time"=>nil}}]]},
                  "SatisfiesAll_E463FEA2_9209_4FA0_AAEF_3466B24A0536_12733BC6_DBCD_4ABD_81D9_BA4775D473AF"=>
                   {"results"=>
                     [[{"id"=>"57169443b0ede6719f002939",
                        "json"=>
                         {"_id"=>"57169443b0ede6719f002939",
                          "admitTime"=>nil,
                          "admitType"=>nil,
                          "codes"=>{"CPT"=>["90834"]},
                          "description"=>"Encounter, Performed: Psych Visit",
                          "diagnosis"=>nil,
                          "dischargeDisposition"=>nil,
                          "dischargeTime"=>nil,
                          "end_time"=>1348152600,
                          "mood_code"=>"EVN",
                          "negationInd"=>nil,
                          "negationReason"=>nil,
                          "oid"=>"2.16.840.1.113883.3.560.1.79",
                          "performer_id"=>nil,
                          "principalDiagnosis"=>
                           {"_id"=>"5776af2eb0ede62f0f000003",
                            "code"=>"2506003",
                            "code_system"=>"SNOMED-CT",
                            "title"=>"Dysthymia"},
                          "reason"=>nil,
                          "specifics"=>nil,
                          "start_time"=>1348151400,
                          "status_code"=>{"HL7 ActStatus"=>["performed"]},
                          "time"=>nil}}],
                      [{"id"=>"57169443b0ede6719f002939",
                        "json"=>
                         {"_id"=>"57169443b0ede6719f002939",
                          "admitTime"=>nil,
                          "admitType"=>nil,
                          "codes"=>{"CPT"=>["90834"]},
                          "description"=>"Encounter, Performed: Psych Visit",
                          "diagnosis"=>nil,
                          "dischargeDisposition"=>nil,
                          "dischargeTime"=>nil,
                          "end_time"=>1348152600,
                          "mood_code"=>"EVN",
                          "negationInd"=>nil,
                          "negationReason"=>nil,
                          "oid"=>"2.16.840.1.113883.3.560.1.79",
                          "performer_id"=>nil,
                          "principalDiagnosis"=>
                           {"_id"=>"5776adedb0ede62ed3000060",
                            "code"=>"2506003",
                            "code_system"=>"SNOMED-CT",
                            "title"=>"Dysthymia"},
                          "reason"=>nil,
                          "specifics"=>nil,
                          "start_time"=>1348151400,
                          "status_code"=>{"HL7 ActStatus"=>["performed"]},
                          "time"=>nil}}]]},
                  "Union_CC7491B7_6A9C_48A7_B828_7AB72D05B203_6F286505_F0D4_4E3F_B8BB_350D9A353988"=>
                   {"results"=>
                     [[{"id"=>"57169443b0ede6719f002939",
                        "json"=>
                         {"_id"=>"57169443b0ede6719f002939",
                          "admitTime"=>nil,
                          "admitType"=>nil,
                          "codes"=>{"CPT"=>["90834"]},
                          "description"=>"Encounter, Performed: Psych Visit",
                          "diagnosis"=>nil,
                          "dischargeDisposition"=>nil,
                          "dischargeTime"=>nil,
                          "end_time"=>1348152600,
                          "mood_code"=>"EVN",
                          "negationInd"=>nil,
                          "negationReason"=>nil,
                          "oid"=>"2.16.840.1.113883.3.560.1.79",
                          "performer_id"=>nil,
                          "principalDiagnosis"=>
                           {"_id"=>"5776af2eb0ede62f0f000003",
                            "code"=>"2506003",
                            "code_system"=>"SNOMED-CT",
                            "title"=>"Dysthymia"},
                          "reason"=>nil,
                          "specifics"=>nil,
                          "start_time"=>1348151400,
                          "status_code"=>{"HL7 ActStatus"=>["performed"]},
                          "time"=>nil}}],
                      [{"id"=>"57169443b0ede6719f002939",
                        "json"=>
                         {"_id"=>"57169443b0ede6719f002939",
                          "admitTime"=>nil,
                          "admitType"=>nil,
                          "codes"=>{"CPT"=>["90834"]},
                          "description"=>"Encounter, Performed: Psych Visit",
                          "diagnosis"=>nil,
                          "dischargeDisposition"=>nil,
                          "dischargeTime"=>nil,
                          "end_time"=>1348152600,
                          "mood_code"=>"EVN",
                          "negationInd"=>nil,
                          "negationReason"=>nil,
                          "oid"=>"2.16.840.1.113883.3.560.1.79",
                          "performer_id"=>nil,
                          "principalDiagnosis"=>
                           {"_id"=>"5776adedb0ede62ed3000060",
                            "code"=>"2506003",
                            "code_system"=>"SNOMED-CT",
                            "title"=>"Dysthymia"},
                          "reason"=>nil,
                          "specifics"=>nil,
                          "start_time"=>1348151400,
                          "status_code"=>{"HL7 ActStatus"=>["performed"]},
                          "time"=>nil}}]]}}}
```

**Comment by edeyoung** [comment]
Looks good other than changing the failure output to be more helpful to the user.

File: lib/tasks/colorization_tests.rake (hunk 2 above)
Comment (pkmitre) [inline lib/tasks/colorization_tests.rake]: This is really targeted at looking at changes to the colorization code rather than to calculation code (there's a different rake task for regressing those... with I suppose a lot of overlap), perhaps it should just include the specificsRationale and leave out the rationale? Looking at a specificsRationale change, it looks more like

```
FAILURE: User kjenkinsbrown@meditech.com measure CMS171v5 population 1 patient 'DENEX_CABG_2EPI D_SCIP1SQL15'

         Difference:

           {"specificsRationale"=>
             {"DENEX"=>
               {"GROUP_satisfiesAny_CHILDREN_261"=>[nil, false],
                "GROUP_UNION_CHILDREN_275"=>[nil, false],
                "precondition_325"=>[nil, false],
                "DENEX"=>[nil, false]}}}
```

It provides a little more data (which population, what the change is). Worth it? Thoughts?
